micronaut-mcp-server-java-sdk/src/main/java/io/micronaut/mcp/server/registry/ResourceTemplateRegistry.java

```diff
@@ -0,0 +1,181 @@
+        String mimeType = method.stringValue(ResourceTemplate.class, MIME_TYPE_PROPERTY).orElse(ResourceTemplate.DEFAULT_MIME_TYPE);
+        return new McpSchema.ResourceTemplate(uri, name, title, description, mimeType, null, null);
+    }
+
+    private <B> BiFunction<McpSyncServerExchange, McpSchema.ReadResourceRequest, McpSchema.ReadResourceResult> syncHandler(
+        BeanDefinition<B> beanDefinition,
+        ExecutableMethod<B, Object> method
+    ) {
+        return (exchange, request) -> invokeAndMap(beanDefinition, method, exchange, request);
+    }
+
+    private <B> BiFunction<McpAsyncServerExchange, McpSchema.ReadResourceRequest, Mono<McpSchema.ReadResourceResult>> asyncHandler(
+        BeanDefinition<B> beanDefinition,
+        ExecutableMethod<B, Object> method
+    ) {
+        return (exchange, request) -> Mono.just(invokeAndMap(beanDefinition, method, exchange, request));
+    }
+
+    private <B> BiFunction<McpTransportContext, McpSchema.ReadResourceRequest, McpSchema.ReadResourceResult> statelessSyncHandler(
+        BeanDefinition<B> beanDefinition,
+        ExecutableMethod<B, Object> method
+    ) {
+        return (ctx, request) -> invokeAndMap(beanDefinition, method, ctx, request);
+    }
+
+    private <B> BiFunction<McpTransportContext, McpSchema.ReadResourceRequest, Mono<McpSchema.ReadResourceResult>> statelessAsyncHandler(
+        BeanDefinition<B> beanDefinition,
+        ExecutableMethod<B, Object> method
+    ) {
+        return (ctx, request) -> Mono.just(invokeAndMap(beanDefinition, method, ctx, request));
+    }
+
+    private <B> McpSchema.ReadResourceResult invokeAndMap(BeanDefinition<B> beanDefinition,
+                                                          ExecutableMethod<B, Object> method,
+                                                          Object mcpTransportContext,
+                                                          McpSchema.ReadResourceRequest request) {
+        B bean = beanContext.getBean(beanDefinition);
+
+        ExecutableBinder<UriTemplateReadResourceRequest> executableBinder = new DefaultExecutableBinder<>(
+            prepareBoundVariables(method, List.of(resolveMcpTransportContext(mcpTransportContext), request)));
+
+        String uriTemplate = method.getAnnotation(ResourceTemplate.class)
+            .stringValue(URI_TEMPLATE_PROPERTY)
+            .orElseThrow(() -> new ConfigurationException("Missing required member uriTemplate in @ResourceTemplate annotation"));
+        BoundExecutable executable = executableBinder.bind(method, argumentBinderRegistry, new UriTemplateReadResourceRequest(uriTemplate, request));
+        Object result = executable.invoke(bean);
+        if (result instanceof McpSchema.ReadResourceResult r) {
+            return r;
+        }
+        if (result instanceof String s) {
+            String mimeType = method.getAnnotation(ResourceTemplate.class)
```

Comment:
this can be `method.stringValue(ResourceTemplate.class, MIME_TYPE_PROPERTY)` which would be more efficient